Add support for generating and configuring repo metadata gpg keys

PaulSD opened the pull request:
See https://bugzilla.redhat.com/show_bug.cgi?id=1410638 and https://github.com/candlepin/subscription-manager/pull/1749

PaulSD commented:
*ping*

ekohl commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,64 @@
+) {
+  if $enable {
+
+    $gpg_key_params = "Key-Type: RSA\nKey-Length:4096\nExpire-Date: 0\n"
```
Could you make this a parameter too? Experience has taught me that at some point someone will want to override this.

ekohl commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,64 @@
+    $fp_cmd = "`${fp_cmd_print} | ${fp_cmd_parse}`"
+    exec { 'extract public repomd gpg key':
+      onlyif  => "/usr/bin/test -e '${dir}'",
+      unless  => "/usr/bin/test -e '${pub_file}'",
```
Would the ```creates``` parameter be the same?

ekohl commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,64 @@
+      command => "/usr/bin/gpg --homedir '${dir}' --armor --export ${fp_cmd} --output '${pub_file}'",
+    }
+
+    file { '/etc/pulp/server/plugins.conf.d/yum_distributor.json':
```
I'd prefer if this module didn't manage pulp resources. We combine them in puppet-katello and puppet-foreman_proxy_content.

ekohl commented on spec/classes/certs_repomd_gpg_spec.rb:
```diff
@@ -0,0 +1,15 @@
+    on_supported_os['redhat-7-x86_64']
+  end
+
+  describe 'with default parameters' do
```
This is with ```enable => false``` but could you add a test where it's set to true?

ekohl commented on templates/rhsm-katello-reconfigure.erb:
```diff
@@ -36,7 +50,8 @@ then
     --rhsm.repo_ca_cert="%(ca_cert_dir)s$KATELLO_SERVER_CA_CERT" \
-    --rhsm.baseurl="$BASEURL"
+    --rhsm.baseurl="$BASEURL" \
+    $REPOMD_GPG_PARAM
```
This can be empty in which case the backslash might do something unexpected. The same applies below

ekohl commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,64 @@
+    $gpg_name_params = "${gpg_name_param}${gpg_comment_param}${gpg_email_param}"
+    $gpg_params = "${gpg_key_params}${gpg_name_params}%no-protection\n%commit\n"
+    if $regenerate {
+      file { $dir:
```
This will lead to a duplicated resource below. I think the only way is an exec. Also ensure it's executed before the directory is created again.

PaulSD commented:
I think I have addressed most of your concerns.

However, I'm not sure how to approach the pulp configuration part.  If we move the pulp configuration to another module, then we will need an additional parameter in that other module to determine whether pulp should be configured to sign repo metadata, and users will need to specify both --certs-enable-repomd-gpg=true and this other parameter to properly configure pulp.  The problem is that pulp will break if it is configured to sign repo metadata but does not have a gpg key, so users who specify this other parameter (to configure pulp) without also specifying --certs-enable-repomd-gpg=true will end up with a broken system.  Leaving the pulp configuration here ensures that pulp is never configured unless a gpg key has been generated.  Given that, do you still want me to move the pulp configuration to another module?

PaulSD commented:
Ok.  I have removed the pulp config.  Thanks!

ekohl commented:
From a puppet code perspective this looks good.

ekohl commented:
That may have been a bit brief. I'll have a look later today. Since we're about to do releases for the katello 3.6 release it'd be great to get this in. For that to happen the puppet-pulp PR needs an update and a PR to puppet-katello (and also puppet-foreman_proxy_content?) to combine the functionality is needed.

PaulSD commented:
I've been working on updating puppet-katello and puppet-foreman_proxy_content ... Now that I see some of the bigger picture, I see that I missed some important details in my changes to puppet-certs here ... But I'm having a hard time following the overall design/flow of these modules as I try to figure out how to implement this properly.  Is there developer documentation for these somewhere?  Or can you help me understand this better?

Specifically:

- My initial assumption was that a single CA is generated on the katello server, the CA cert is added to a single RPM that is pushed out to all clients, then keys/certs for each proxy are generated on the katello server and pushed out to each proxy.  However, I see some code paths that seem to contradict my expectations ... For example, it looks like certs::katello is called to generate the client RPM on each proxy?  Do clients get different certs RPMs depending on which proxy they connect to?  Am I totally misunderstanding how this works?
- yum doesn't handle GPG subkeys properly, so assuming all clients are supposed to get the same CA cert regardless of which Pulp they connect to, we must either use the same private GPG key to sign metadata on every Pulp, or we must generate a separate private GPG key for every Pulp and distribute all of the associated public keys to every client.  Since the existing process is designed to distribute certs to the clients only once (and not re-distribute the certs each time a new Pulp is added), I think the only practical option is to generate a single private GPG key and distribute it to every Pulp.
- So, the first step is to generate the GPG key at approximately the same time as the CA is generated by puppet-certs.  It looks like my changes to manifests/init.pp in this PR are the appropriate way to do this, and I see where this is called by puppet-katello/manifests/application.pp, so this much makes sense to me.  I still need to update puppet-katello with the new arguments for puppet-certs and puppet-pulp, but this should be straightforward.
- Next, the public GPG key needs to be added to the certs RPM that will be installed on clients.  My changes to manifests/katello.pp in this PR handle that.
  - However, I only see where certs::katello is called on the foreman proxy servers ... which doesn't really make sense to me as mentioned above ... Where is it called on the katello server, and why is it called at all on the foreman proxy servers?
- After that, the private GPG key needs to be added to the certs tarball that is distributed to each foreman proxy.  This part isn't implemented at all yet in this PR.
  - It looks like I need to update manifests/foreman_proxy_content.pp to make this happen.  However, I don't see where/when/how certs::foreman_proxy_content is called?
  - Once it is called, manifests/foreman_proxy_content.pp calls a bunch of other certs::* classes.  All of those classes have $deploy arguments that don't quite make sense to me, especially since I don't see anywhere that $deploy is set to anything other than the default value.  Are these classes called with $deploy=false on the katello server and $deploy=true on each foreman proxy, or something like that?  I'm expecting the certs to be generated on the katello server and deployed on each proxy, but I'm not following how that is happening in the code...
  - manifests/foreman_proxy_content.pp then calls manifests/tar_create.pp, which tars up RPMs from under /root/ssl-build/ ... It looks like those RPMs are generated by `certs::keypair`?  Do I need to write something to generate an RPM containing the GPG key, or can I just roll the GPG key directly into this tarball?
- Finally, the private GPG key needs to be installed on each foreman proxy.  This is also not implemented yet in this PR.
  - I see that puppet-foreman_proxy_content/manifests/init.pp calls certs::tar_extract before calling the `certs` class, but I don't see where the RPMs in the tar are installed?  Are they automatically extracted by the `cert` or `certs::keypair` provider?

ehelms commented:
On Sat, Jan 27, 2018 at 11:37 AM, Paul Donohue <notifications@github.com>
wrote:

> I've been working on updating puppet-katello and
> puppet-foreman_proxy_content ... Now that I see some of the bigger picture,
> I see that I missed some important details in my changes to puppet-certs
> here ... But I'm having a hard time following the overall design/flow of
> these modules as I try to figure out how to implement this properly. Is
> there developer documentation for these somewhere? Or can you help me
> understand this better?
>
> Specifically:
>
>    - My initial assumption was that a single CA is generated on the
>    katello server, the CA cert is added to a single RPM that is pushed out to
>    all clients, then keys/certs for each proxy are generated on the katello
>    server and pushed out to each proxy. However, I see some code paths that
>    seem to contradict my expectations ... For example, it looks like
>    certs::katello is called to generate the client RPM on each proxy? Do
>    clients get different certs RPMs depending on which proxy they connect to?
>    Am I totally misunderstanding how this works?
>
> You are spot on. Each proxy is deployed with it's own consumer RPM. A
client wishing to connect to a given proxy and use it as an "isolated"
endpoint must use the consumer RPM on the proxy. This allows a client to
talk only to a proxy and thus allow the proxy to serve as a single gateway
in a datacenter for example. The reason for different RPMs is that RHSM has
to be configured differently depending on the endpoint the client is
talking to.

>
>    - yum doesn't handle GPG subkeys properly, so assuming all clients are
>    supposed to get the same CA cert regardless of which Pulp they connect to,
>    we must either use the same private GPG key to sign metadata on every Pulp,
>    or we must generate a separate private GPG key for every Pulp and
>    distribute all of the associated public keys to every client. Since the
>    existing process is designed to distribute certs to the clients only once
>    (and not re-distribute the certs each time a new Pulp is added), I think
>    the only practical option is to generate a single private GPG key and
>    distribute it to every Pulp.
>
> I cannot unfortunately address the technical aspects you've outlined here
yet because I need to understand better why this change is needed. I'll
outline the GPG key and repository workflow as I understand it as a Katello
developer and that might help you point out to me what workflow I am
missing that this solves.

I have an external yum repository A. The packages in yum repository A are
signed by gpg key "repo-a-key". I create a Katello repository object
pointing to repository A. I then create a GPG key object in Katello with
the contents of "repo-a-key". I assign "repo-a-key" to the Katello
repository object. I then sync repository A and assign it to my content
host. My content host now has access to repository A served from Katello
and gets the GPG key for the repository from the Katello server.

So given that workflow, can you outline what workflow you are trying to
solve?

>
>    - So, the first step is to generate the GPG key at approximately the
>    same time as the CA is generated by puppet-certs. It looks like my changes
>    to manifests/init.pp in this PR are the appropriate way to do this, and I
>    see where this is called by puppet-katello/manifests/application.pp,
>    so this much makes sense to me. I still need to update puppet-katello with
>    the new arguments for puppet-certs and puppet-pulp, but this should be
>    straightforward.
>    - Next, the public GPG key needs to be added to the certs RPM that
>    will be installed on clients. My changes to manifests/katello.pp in this PR
>    handle that.
>       - However, I only see where certs::katello is called on the foreman
>       proxy servers ... which doesn't really make sense to me as mentioned above
>       ... Where is it called on the katello server, and why is it called at all
>       on the foreman proxy servers?
>    - After that, the private GPG key needs to be added to the certs
>    tarball that is distributed to each foreman proxy. This part isn't
>    implemented at all yet in this PR.
>       - It looks like I need to update manifests/foreman_proxy_content.pp
>       to make this happen. However, I don't see where/when/how
>       certs::foreman_proxy_content is called?
>       - Once it is called, manifests/foreman_proxy_content.pp calls a
>       bunch of other certs::* classes. All of those classes have $deploy
>       arguments that don't quite make sense to me, especially since I don't see
>       anywhere that $deploy is set to anything other than the default value. Are
>       these classes called with $deploy=false on the katello server and
>       $deploy=true on each foreman proxy, or something like that? I'm expecting
>       the certs to be generated on the katello server and deployed on each proxy,
>       but I'm not following how that is happening in the code...
>       - manifests/foreman_proxy_content.pp then calls
>       manifests/tar_create.pp, which tars up RPMs from under /root/ssl-build/ ...
>       It looks like those RPMs are generated by certs::keypair? Do I need
>       to write something to generate an RPM containing the GPG key, or can I just
>       roll the GPG key directly into this tarball?
>    - Finally, the private GPG key needs to be installed on each foreman
>    proxy. This is also not implemented yet in this PR.
>       - I see that puppet-foreman_proxy_content/manifests/init.pp calls
>       certs::tar_extract before calling the certs class, but I don't see
>       where the RPMs in the tar are installed? Are they automatically extracted
>       by the cert or certs::keypair provider?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/puppet-certs/pull/188#issuecomment-360996595>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAR581P6-sdV_FMl2JejDweuMWpvT0yIks5tO1CsgaJpZM4RLt6W>
> .
>



-- 
Eric D. Helms
Red Hat Engineering

PaulSD commented:
Thanks for the reply!

> I cannot unfortunately address the technical aspects you've outlined here
> yet because I need to understand better why this change is needed. I'll
> outline the GPG key and repository workflow as I understand it as a Katello
> developer and that might help you point out to me what workflow I am
> missing that this solves.

There are actually two GPG keys for each repository:
1) The GPG key used to sign the RPMs in the repository
2) The GPG key used to sign the repository's repomd.xml file
The same GPG key could be used for both of these, or two different keys could be used for them.

Your workflow describes the handling of the first GPG key.  Since Pulp doesn't modify the RPMs themselves, the GPG signatures on the RPMs do not need to be modified, so the same public GPG key can be used by clients of Pulp as is used by clients which use the upstream repository directly.

However, Pulp generates a repomd.xml file that is different than the one published by the upstream repository.  So, clients of Pulp must use a different public GPG key to verify the GPG signature on that file than clients which use the upstream repository directly.  The private portion of this GPG key must live on the Pulp server, and the public portion of it must be distributed to clients as part of the process that configures yum on the client to connect to Pulp.

I think I have the simple case worked out - if there is only one Katello/Pulp server and there are no proxies.  In its current form, this PR will generate a GPG key on the Pulp server, will add the public portion of the GPG key to the RPM that installs keys on clients, and https://github.com/theforeman/puppet-pulp/pull/302 will configure Pulp to use the generated GPG key to sign repomd.xml files.  Except for a little bit of glue in puppet-katello and puppet-foreman_proxy_content to tie it all together, I think this is pretty much all that is needed.

However, I'm stuck on the case where proxies are involved ... I'm trying to more or less emulate the existing CA handling behavior, but I'm not understanding the code flow when proxies are involved.  And to make things more complicated, unlike X509 certs issued by a CA, it's not really possible to generate multiple GPG keys that are all trusted by a single "CA" public key that can be distributed to all clients.  So, I guess the first questions I'm trying to answer here are:
Should I use one GPG key for all proxies (private key shared across all proxies), and distribute the public portion of that one key to all clients?  Or should I use a different GPG key on every proxy?  And if I use a different GPG key on every proxy, then should I distribute all public GPG keys for all proxies to all clients, or only distribute one public GPG key for one proxy to each client (preventing clients from connecting to any other proxies)?

ehelms commented:
I did some digging of my own and based on your solid write-up, Pulp docs and my Katello knowledge I have a few thoughts and recommendations. Unfortunately, some of these recommendations will require more work depending on where we land.

The first thing as I understand it on how Pulp works, they require a private GPG key on the server and every Capsule and support only one key for an entire installation of Pulp to sign the metadata which is different than how RPM signing is handled and can support 1-N GPG keys. So in this sense, there can only be one GPG key for the entire installation for metadata (which is me re-iterating it for my own sanity).

The first thought is you linked to puppet-pulp change that makes it so that *every* repository in your entire Katello/Pulp will have metadata signing. That seems like an extreme case, that may work for you but I am not sure if it would make sense for all user workflows. So a thought would be to add this to Katello repositories as a configuration option that would allow individual repositories to be configured for metadata signing.

The second thought is that the `puppet-pulp` module should support accepting a GPG key and configuring it properly for use. This `puppet-certs` module could create and define that key though.

The third thought goes to your actual question. If you stick strictly to the idea that a client is only registered and getting content from a single proxy, then if the client ever switches proxies they have to change client RPMs that are installed. Now, if you go outside that workflow, or use http repositories across proxies you'd have to distribute all the GPG public keys. So I lean towards a single GPG key across the infrastructure.

@jlsherrill Any thoughts in this area?

@PaulSD I don't think you need to go back to yum to add support, but I'd be curious if DNF has support and if it would be worth investing in giving it the support but not a blocker.

ehelms commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,60 @@
+    $pub_file = "${dir}/pub.gpg"
+    $fp_cmd_print = "/usr/bin/gpg --homedir '${dir}' --list-secret-keys --with-colons"
+    $fp_cmd_parse = "/usr/bin/awk -F: '\$1 == \"fpr\" {print \$10; exit;}'"
+    $fp_cmd = "`${fp_cmd_print} | ${fp_cmd_parse}`"
```
What is `fp`? A descriptive name might help with readability.

ehelms commented on manifests/repomd_gpg.pp:
```diff
@@ -0,0 +1,60 @@
+
+    $pub_file = "${dir}/pub.gpg"
+    $fp_cmd_print = "/usr/bin/gpg --homedir '${dir}' --list-secret-keys --with-colons"
+    $fp_cmd_parse = "/usr/bin/awk -F: '\$1 == \"fpr\" {print \$10; exit;}'"
```
Are we guaranteed awk and gpg exist on every system? Or do we need some package dependecies?

ehelms commented:
@PaulSD Theoritically this can be split into two pieces right:

 1) The generation of a GPG private and public key where are generic operations
 2) Configuring RHSM script for repomd_gpg that uses a GPG key

If I am correct there, then we could make the gpg key aspects generic and let the RHSM manifest either generate a key or take one as input?

ehelms commented:
@PaulSD checking back with you on this since its been a few months

PaulSD commented:
Oy ... time flies ...

I will try to pick this back up.

PaulSD commented:
FYI: I am working on this, but it is taking a while because I'm doing this as a background task and I've been slowly struggling through setting up a testing environment for a capsule using forklift (which hasn't gone as smoothly as I had hoped).

PaulSD commented:
To answer the last of my earlier questions that isn't answered in my previous comment:

A single CA is generated on the primary server and is copied to each capsule.  Each server (primary and capsule) generates its own client bootstrap RPM, and the KATELLO_SERVER variable in this RPM is unique to the server.  However, all of these RPMs include the same certificate (the single CA).  Clients do not get different certs from different servers; they only get different KATELLO_SERVER settings.

This means that clients can currently switch between servers simply by changing the server name used in their configuration.  If we were to use a different GPG key for each server, then we would loose this capability, and users would be required to install a different bootstrap RPM when switching servers.  So I'm still leaning toward using a single GPG key across all servers.

PaulSD commented:
To respond to @ehelms in https://github.com/theforeman/puppet-certs/pull/188#issuecomment-367890525 : 

> The first thought is you linked to puppet-pulp change that makes it so that every repository in your entire Katello/Pulp will have metadata signing. That seems like an extreme case, that may work for you but I am not sure if it would make sense for all user workflows. So a thought would be to add this to Katello repositories as a configuration option that would allow individual repositories to be configured for metadata signing.

The problem here is that the relevant setting is global in Pulp itself.  Pulp does not have a way to enable this on a per-repository basis - either all repositories are signed or none are.
However, even if it did support per-repository configuration ... Signed repositories are fully backward compatible with clients that don't do signature checks ... And if a client supports checking metadata signatures on any repository on a server then it will inherently support checking signatures on every repository on that server ... So is there any case where a specific repository would need to be unsigned?  Furthermore, the security benefits of enabling repository signatures are greatly diminished if any repository is unsigned ... So is there any case where it would be necessary to sign one repository but acceptable to have another unsigned repository?

> The second thought is that the puppet-pulp module should support accepting a GPG key and configuring it properly for use. This puppet-certs module could create and define that key though.

The problem here is that it is not possible to pass around a private GPG key like you would pass around a private X509 key.  Private GPG keys live permanently in a GPG keyring, they cannot be exported or imported.  You can copy or move the entire GPG keyring directory, but that's about it.  In addition, it is rather difficult to reliably "configure" a GPG keyring after it has been created; it really needs to be configured at creation time.

At best, puppet-certs could generate and configure a keyring in a temporary location, then pass the name of the keyring directory to puppet-pulp and the RHSM script so they can move/copy the whole directory to an appropriate location.  Would that buy us anything?

PaulSD commented:
Ok, so my current plan is:
- Move my new [repomd_gpg arguments](https://github.com/theforeman/puppet-certs/pull/188/commits/a374936c539ae7cd9115a675053cae372686ca0d#diff-60ae41fd0a31977447947f59940ee9a4) to puppet-foreman_proxy_content
  - Rework the arguments to make sense when generating a GPG key on the primary server and deploying a generated GPG key to a capsule
  - Pass the arguments through to puppet-pulp
- Add relevant arguments to puppet-pulp and add calls to puppet-certs to actually generate/deploy the GPG key
- Rework this PR in puppet-certs
  - Make GPG key generation generic ... puppet-pulp will call it with relevant repomd_gpg specific arguments
  - Properly handle the $generate and $deploy flags, and mimic the existing behavior (generate the key under /root/, generate an RPM containing the key, and install the RPM to deploy)
  - If necessary, update the tar_create and tar_extract manifests to support the new RPM
  - If time permits, add optional support for generating GPG subkeys (maybe we can use them in the future)
- Modify katello-certs-tools to support generating GPG key RPMs
  - I can generate the GPG keys themselves in puppet, so I don't need to implement that part in katello-certs-tools ... Should I implement GPG key generation in katello-certs-tools anyway just for consistency with the SSL key generation stuff, or is it OK for me to do things a little differently than SSL and generate GPG keys in puppet and only use katello-certs-tools to build the RPM?
- Modify puppet-pulp (or do I need to modify the existing certs::katello call in puppet-foreman_proxy_content) to pass the public GPG key to certs::katello so it can be added to the client bootstrap RPM

PaulSD commented:
Ping @ekohl @ehelms 
This is ready for re-review.

sean797 commented on README.md:
```diff
@@ -78,7 +82,7 @@ cert-specific resources:
 `katello_ssl_tool`. It works as follow:
 
-1. **generation** - the artefact of this phase is an RPM with the keys
+1. **generation** - the artifact of this phase is an RPM with the keys
```
I'm fairly sure `artefact` is correct.

PaulSD replied:
https://www.merriam-webster.com/dictionary/artefact
*shrug*

sean797 commented on manifests/foreman_proxy_content.pp:
```diff
@@ -30,7 +33,14 @@
+  } else {
+    class { '::certs::repomd_gpg': hostname => $foreman_proxy_fqdn }
+    certs::tar_create { $certs_tar:
+      subscribe => Class['certs::puppet', 'certs::foreman', 'certs::foreman_proxy', 'certs::qpid', 'certs::qpid_router', 'certs::apache', 'certs::qpid_client', 'certs::repomd_gpg'],
```
You can build the `subscribe` list dynamically as I have in another PR https://github.com/theforeman/puppet-certs/pull/199/files#diff-1aca67488955df8f44fd9c808ade15f1R31

PaulSD replied:
Ok, I'll change this.

sean797 commented on manifests/init.pp:
```diff
@@ -67,6 +67,58 @@
+# $repomd_gpg_name::             Name attribute for the GPG key generated for
+#                                repository metadata signing
+#
+# $repomd_gpg_comment::          Comment attribute for the GPG key generated
```
I'd suggest some of there should definitely be advanced options, or configurable via hiera.

PaulSD replied:
They are all advanced options.  I didn't add anything to the non-advanced parameter list.

sean797 commented on lib/puppet/provider/gpg/gpg.rb:
```diff
@@ -0,0 +1,318 @@
+    end
+    rpm_file = latest_rpm
+    if rpm_file
+      rpm('-Uvh', '--force', rpm_file)
```
Do we really need `--force` ?

PaulSD replied:
Not sure ... I copied that code from https://github.com/theforeman/puppet-certs/blob/master/lib/puppet/provider/katello_ssl_tool.rb#L77

PaulSD commented:
A complete acceptance test (covering all of the relevant PRs) is to install a Foreman server or capsule, install the generated client bootstrap RPM on a client, verify that `repo_gpgcheck` is set to 1 in yum.conf on the client, then verify that yum on the client can successfully install an RPM.

PaulSD commented:
I don't think these new test failures are caused by my changes.

PaulSD commented:
[test]

PaulSD commented:
Ping @ekohl @ehelms @sean797 

I think I addressed / responded to all of the comments back in July, but I haven't heard anything since.

I just updated this PR to fix some merge conflicts ... These test failures are definitely not related to my changes.  Looks like a dependency conflict.

ehelms commented:
The PRs related to this for Pulp 2 have been closed as we removed Pulp 2 from deployments. There are parts of this PR that have moved (such as RHSM reconfigure script) to other modules. Aspects of this look generic enough to be potentially be relevant to a Pulp 3 solution. We can therefore opt to leave this PR open as we explore the Pulp 3 implementation or close and approach it fresh in Pulp 3.

ehelms commented:
I am going to opt to close this in favor of seeing how the Pulp 3 design shakes out.
